ARCHITECTURE.md

```diff
@@ -0,0 +1,207 @@
+  * Rails has been a great long-term fit to our problem space and we benefit from continuing to get upgrades as Rails has improved. Doing it “their way” makes upgrades happen semi-automatically.
+
+</details>
+
+### Tenet 6: Managed Services that implement an Open Source Protocol / API (vs proprietary managed services or self-hosted open source services)
+
+When making technical selections, prefer managed services that implement an Open Protocol or an Open Source API. For example, we don’t host our own MySQL instance, but we ALSO don’t use ProprietarySQLDB(TM). Instead, we use a 3rd party managed service that implements the MySQL protocol.
+
+<details>
+  <summary>Why?</summary>
+
+  * Gives us the best of both worlds:
+    * Self-hosting many open source servers, especially stateful services like DBs, requires ongoing infra admin work and monitoring attention. We have a small team.  
+    * Developing against an API with an Open Source implementation gives us flexibility:  
+      * …to implement a “local dev” for the service using the open source version  
+      * …to change vendors if need be (cost, availability, in-kind donations, etc)  
+
+</details>
+
+### Tenet 7: Use REST (vs. GraphQL) for APIs
+
+As we add APIs for our platform, we will publish them using REST. Client-side apps should communicate with the server using REST endpoints.
+
+<details>
+  <summary>Why?</summary>
+
+  * KISS: REST has been a simple and an effective match to our needs.  
+  * Our labs and other fat client-side JS apps don’t tend to have the “complex aggregation of dynamic data from many sources” property that benefits from more complex systems like GraphQL.  
+  * Memoization is easily applied to REST-based calls (good for hosting cost and at-scale events like hour of code).
+
+</details>
+
+### Tenet 8: Containers (vs. Chef/scripts for deployments)
+
+We will move towards deploying our code using containers over further investing in other deployment mechanisms (either Chef or custom scripts).
+
+<details>
+  <summary>Why?</summary>
+
+  * Provides a common deployment and development pattern across our engineering teams.  
+  * Decreases the work required to port our services between cloud providers  
+  * Easier to build a local dev environment that effectively mimics prod
+
+</details>
+
+<details>
+  <summary>Exceptions</summary>
+
+  * Any storage layer in production (e.g., MySQL/RDS/S3) should not be containerized.  
+  * We will still rely on vendor-specific templates for infrastructure configuration (e.g., Cloud Formation)  
+  * Scripts can still be helpful to make complex multi-step operations easier/clearer
+
+</details>
+
+---
+
+### ⬆️ Above ⬆️ this line live the architecutral tenets. They should be kept at the top of the file and edited sparingly.
+
+### ⬇️ Below ⬇️ this line is general architecture information: please edit freely!
+
+---
+
+# Code.org's Overall Architecture...
+
+Overview of our architecture goes here, please fill me in!
```

Comment:
I'm lukewarm on including this because it feels like could quickly become a dumping ground. I'm guessing your goal is to keep non-tenets out of the tenets section above, which I support, but would like to consider some alternatives.

for example: perhaps just say "further documentation of current architecture can be found in Google Drive > Shared drives > Engineering > Architecture"?

Reply:
I think an /ARCHITECTURE.md doc will be at permanent "risk" for somebody deciding to document, well, our architecture there. I'm not trying to encourage a dumping ground here (although personally I would prefer we actually had normal ARCHITECTURE.md material here vs google drive), but I'd rather have a dumping ground below the line than not "defensively" anticipate a possible way this goes off the rails.